### PR TITLE
Add support for idobao id80 iso layout

### PR DESCRIPTION
**Pull request by cwr10010**
## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [X] Keyboard (addition or update)
- [X] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [X] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [X] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [X] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [X] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Review by drashna — COMMENTED** [review]
The only change here is adding an ISO layout, and nothing else?  

If so, I'm not really a fan of splitting this up into two different revisions.  Rather `LAYOUT_ansi` and `LAYOUT_iso` in the keyboard will work.  Additionally, VIA is capable of layout grouping, so you can toggle between the two, without problem.

**Comment by cwr10010** [comment]
> The only change here is adding an ISO layout, and nothing else?
> 
> If so, I'm not really a fan of splitting this up into two different revisions. Rather `LAYOUT_ansi` and `LAYOUT_iso` in the keyboard will work. Additionally, VIA is capable of layout grouping, so you can toggle between the two, without problem.

Yes that is the only change. In fact the board is a hot swap board with two revisions that are wired differently. You may want to have  look at the Enter key that is on the A-row for ansi and on the Q row for ISO. If your suggestion is that the board would support ISO just in VIA this would not follow my intention of this PR.

**Comment by drashna** [comment]
In that case, a `LAYOUT_all` for via/etc and a layout for ansi and iso would take care the issue, as the layout allows you to remap the positions.

**Comment by cwr10010** [comment]
This would (imho) break the ability to flash the VIA variant to the board and have it working with ISO out of the box without remapping anything.

**Comment by fauxpark** [comment]
That's exactly what `LAYOUT_all` is for and why we recommend VIA keymaps use it, as it contains all possible matrix positions that can be mapped to the correct keycodes. This way, it doesn't matter whether you have an ANSI or ISO layout, the VIA keymap will work as you expect.

**Comment by cwr10010** [comment]
Then you will have to help me on this one. On the ANSI variant Enter is on Pin 83 and Backslash on Pin 96. With ISO it is the other way around. As what I have understood of VIA it is a layout editor that displays the layout and makes you reassign keys. How can VIA detect if it is ANSI or ISO when it is the same layout mapping, the same firmware, with the same vendor and product id? If I have flashed the via variant my enter key on the ISO board would be on Backslash and I would require to go to via and change that there, right?

**Comment by fauxpark** [comment]
Oh, I did not see your comment above:

> In fact the board is a hot swap board with two revisions that are wired differently.

If the ANSI and ISO layouts are two different boards, then that is fine.

**Comment by cwr10010** [comment]
Just to clarify my understanding on how all works together:

### id80.h
Maps the wiring of the PCB to the physical location on the board. ANSI and ISO have just one main difference here. ISO is having the split left shift in adding pin 11 in the key layout (being KC_NO on ANSI)

### keyboard.c
Maps the physical layout of the board to the keymaps. Here ANSI and ISO are not compatible. ANSI is having KC_ENT on pin 83 and KC_BSPC on pin 96. ISO is the other way around as the routing of the wires is different Physically the Enter key belongs to the A-row on ANSi and to the Q-row on ISO. 

If I understand this right, it is not possible to have VIA on this board and have it working out of the box for both variants. You would need to go to VIA first and reassign the key mapping for the enter key. Based on this I don't know if this is a good idea that one of the variants will not work correctly directly after flashing.

By the way I have a JSON file her ready for a PR for VIA that adds the left split shift key and the different mapping for Enter. via two switches.

File: keyboards/id80/id80.h [hunk 1]
```diff
@@ -35,3 +35,22 @@
     { K07,   K17,   K27,   K37,   K47,   K57,   K67,   K77,   K87,   K97,   KA7   }, \
     { K08,   K18,   K28,   K38,   K48,   K58,   K68,   K78,   K88,   K98,   KC_NO }, \
 }
+
+#define LAYOUT_iso( \
+    K50,  K51, K52, K53, K54,  K55, K56, K57, K58,  KA7, KA5, KA4, KA3,   KA6,  KA2,  \
+    K40, K41, K42, K43, K44, K45, K46, K47, K48, K98, K97, K95, K94, K96,       K92,  \
+    K30,   K31, K32, K33, K34, K35, K36, K37, K38, K88, K87, K85, K84, K83,     K82,  \
+    K20,    K21, K22, K23, K24, K25, K26, K27, K28, K78, K77, K75, K74,               \
```
Comment (noroadsleft) [inline keyboards/id80/id80.h:43]: QMK custom is that Enter key keycodes should be toward the end of the home row.

(1/9)

```suggestion
    K30,   K31, K32, K33, K34, K35, K36, K37, K38, K88, K87, K85, K84,          K82,  \
    K20,    K21, K22, K23, K24, K25, K26, K27, K28, K78, K77, K75, K74, K83,          \
```

File: keyboards/id80/info.json [hunk 2]
```diff
@@ -8,6 +8,98 @@
...
+                {"label":"R", "x":4.5, "y":2.25}, 
+                {"label":"T", "x":5.5, "y":2.25}, 
+                {"label":"Y", "x":6.5, "y":2.25}, 
+                {"label":"U", "x":7.5, "y":2.25}, 
+                {"label":"I", "x":8.5, "y":2.25}, 
+                {"label":"O", "x":9.5, "y":2.25}, 
+                {"label":"P", "x":10.5, "y":2.25}, 
+                {"label":"[", "x":11.5, "y":2.25}, 
+                {"label":"]", "x":12.5, "y":2.25}, 
+                {"label":"Enter", "x":13.5, "y":2.25, "w":1.5, "h":2},
```
Comment (noroadsleft) [inline keyboards/id80/info.json:60]: (2/9)

```suggestion
```

File: keyboards/id80/info.json [hunk 3]
```diff
@@ -8,6 +8,98 @@
...
+                {"label":"F", "x":4.75, "y":3.25},
+                {"label":"G", "x":5.75, "y":3.25},
+                {"label":"H", "x":6.75, "y":3.25},
+                {"label":"J", "x":7.75, "y":3.25},
+                {"label":"K", "x":8.75, "y":3.25},
+                {"label":"L", "x":9.75, "y":3.25},
+                {"label":";", "x":10.75, "y":3.25},
+                {"label":"'", "x":11.75, "y":3.25},
+                {"label":"ISO #", "x":12.75, "y":3.25},
+
```
Comment (noroadsleft) [inline keyboards/id80/info.json:76]: Move the Enter key's object to match the updated `id80.h`.

(3/9)

```suggestion
                {"label":"Enter", "x":13.5, "y":2.25, "w":1.5, "h":2},

```

File: keyboards/id80/iso/keymaps/default/keymap.c [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include QMK_KEYBOARD_H
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+    [0] = LAYOUT_iso(
+        KC_ESC,    KC_F1,   KC_F2,   KC_F3,   KC_F4,     KC_F5,   KC_F6,   KC_F7,   KC_F8,     KC_F9,   KC_F10,  KC_F11,  KC_F12,     MO(1),     KC_INS,
+        KC_GRV,  KC_1,    KC_2,    KC_3,    KC_4,    KC_5,    KC_6,    KC_7,    KC_8,    KC_9,    KC_0,    KC_MINS, KC_EQL,  KC_BSPC,            KC_HOME,
+        KC_TAB,      KC_Q,    KC_W,    KC_E,    KC_R,    KC_T,    KC_Y,    KC_U,    KC_I,    KC_O,    KC_P,    KC_LBRC, KC_RBRC, KC_ENT,         KC_DEL,
+        KC_CAPS,       KC_A,    KC_S,    KC_D,    KC_F,    KC_G,    KC_H,    KC_J,    KC_K,    KC_L,    KC_SCLN,  KC_QUOT,  KC_BSLS,
```
Comment (noroadsleft) [inline keyboards/id80/iso/keymaps/default/keymap.c:23]: (4/9)

```suggestion
        KC_TAB,      KC_Q,    KC_W,    KC_E,    KC_R,    KC_T,    KC_Y,    KC_U,    KC_I,    KC_O,    KC_P,    KC_LBRC, KC_RBRC,                 KC_DEL,
        KC_CAPS,       KC_A,    KC_S,    KC_D,    KC_F,    KC_G,    KC_H,    KC_J,    KC_K,    KC_L,    KC_SCLN,  KC_QUOT,  KC_BSLS, KC_ENT,
```

File: keyboards/id80/iso/keymaps/default/keymap.c [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+        KC_LSFT,   KC_NUBS,  KC_Z,    KC_X,   KC_C,    KC_V,    KC_B,    KC_N,    KC_M,    KC_COMM, KC_DOT,  KC_SLSH, KC_RSFT,          KC_UP,
+        KC_LCTL,   KC_LGUI,   KC_LALT,                       KC_SPC,                              KC_RALT,     KC_RCTL,        KC_LEFT, KC_DOWN, KC_RGHT
+    ),
+    [1] = LAYOUT_iso(
+        RESET,     _______, _______, _______, _______,   _______, _______, _______, _______,   _______, _______, _______, _______,    _______,   KC_MUTE,
+        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,            _______,
+        _______,     RGB_TOG, _______, RGB_MOD, RGB_HUI, RGB_HUD, RGB_SAI, RGB_SAD, RGB_VAI, RGB_VAD, _______, _______, _______, _______,        _______,
+        _______,       _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
```
Comment (noroadsleft) [inline keyboards/id80/iso/keymaps/default/keymap.c:31]: (5/9)

```suggestion
        _______,     RGB_TOG, _______, RGB_MOD, RGB_HUI, RGB_HUD, RGB_SAI, RGB_SAD, RGB_VAI, RGB_VAD, _______, _______, _______,                 _______,
        _______,       _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
```

File: keyboards/id80/iso/keymaps/via/keymap.c [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+    * │ CTRL  │ FN   │ ALT  │                 SPC                  │  ALT  │  CTRL │┌─────┼─────┼─────┤
+    * ├───────┴──────┴──────┴──────────────────────────────────────┴───────┴───────┘│ LFT │ DWN │ RGT │
+    * └─────────────────────────────────────────────────────────────────────────────┴─────┴─────┴─────┘
+    */
+
+    [_BL] = LAYOUT_iso(
+        KC_ESC,  KC_F1,   KC_F2,   KC_F3,   KC_F4,   KC_F5,   KC_F6,   KC_F7,   KC_F8,   KC_F9,   KC_F10,  KC_F11,  KC_F12,  KC_HOME, KC_END,
+        KC_GRV,  KC_1,    KC_2,    KC_3,    KC_4,    KC_5,    KC_6,    KC_7,    KC_8,    KC_9,    KC_0,    KC_MINS, KC_EQL,  KC_BSPC, KC_PGUP,
+        KC_TAB,  KC_Q,    KC_W,    KC_E,    KC_R,    KC_T,    KC_Y,    KC_U,    KC_I,    KC_O,    KC_P,    KC_LBRC, KC_RBRC, KC_ENT,  KC_PGDN,
+        KC_CAPS, KC_A,    KC_S,    KC_D,    KC_F,    KC_G,    KC_H,    KC_J,    KC_K,    KC_L,    KC_SCLN, KC_QUOT, KC_BSLS,
```
Comment (noroadsleft) [inline keyboards/id80/iso/keymaps/via/keymap.c:50]: (6/9)

```suggestion
        KC_TAB,  KC_Q,    KC_W,    KC_E,    KC_R,    KC_T,    KC_Y,    KC_U,    KC_I,    KC_O,    KC_P,    KC_LBRC, KC_RBRC,          KC_PGDN,
        KC_CAPS, KC_A,    KC_S,    KC_D,    KC_F,    KC_G,    KC_H,    KC_J,    KC_K,    KC_L,    KC_SCLN, KC_QUOT, KC_BSLS, KC_ENT,
```

File: keyboards/id80/iso/keymaps/via/keymap.c [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+    * │       │      │ SUP  │                                      │  SUP  │       │┌─────┼─────┼─────┤
+    * ├───────┴──────┴──────┴──────────────────────────────────────┴───────┴───────┘│ BLT │ BLD │ BLS │
+    * └─────────────────────────────────────────────────────────────────────────────┴─────┴─────┴─────┘
+    */
+
+    [_FN] = LAYOUT_iso(
+        RESET,   KC_BRID, KC_BRIU, _______, BL_TOGG, BL_DEC,  BL_INC,  KC_MRWD, KC_MPLY, KC_MFFD, KC_MUTE, KC_VOLD, KC_VOLU, KC_PAUS, KC_PSCR,    
+        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, KC_INS,
+        _______, RGB_TOG, RGB_MOD, RGB_HUI, RGB_HUD, RGB_SAI, RGB_SAD, RGB_VAI, RGB_VAD, _______, _______, _______, _______, _______, KC_DEL,
+        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
```
Comment (noroadsleft) [inline keyboards/id80/iso/keymaps/via/keymap.c:77]: (7/9)

```suggestion
        _______, RGB_TOG, RGB_MOD, RGB_HUI, RGB_HUD, RGB_SAI, RGB_SAD, RGB_VAI, RGB_VAD, _______, _______, _______, _______,          KC_DEL,
        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
```

File: keyboards/id80/iso/keymaps/via/keymap.c [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+    * │       │      │      │                                      │       │       │┌─────┼─────┼─────┤
+    * ├───────┴──────┴──────┴──────────────────────────────────────┴───────┴───────┘│     │     │     │
+    * └─────────────────────────────────────────────────────────────────────────────┴─────┴─────┴─────┘
+    */
+
+    [_E1] = LAYOUT_iso(
+        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
+        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
+        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
+        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
```
Comment (noroadsleft) [inline keyboards/id80/iso/keymaps/via/keymap.c:104]: (8/9)

```suggestion
        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,          _______,
        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
```

File: keyboards/id80/iso/keymaps/via/keymap.c [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+    * │       │      │      │                                      │       │       │┌─────┼─────┼─────┤
+    * ├───────┴──────┴──────┴──────────────────────────────────────┴───────┴───────┘│     │     │     │
+    * └─────────────────────────────────────────────────────────────────────────────┴─────┴─────┴─────┘
+    */
+
+    [_E2] = LAYOUT_iso(
+        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
+        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
+        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
+        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
```
Comment (noroadsleft) [inline keyboards/id80/iso/keymaps/via/keymap.c:131]: (9/9)

```suggestion
        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,          _______,
        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
```

**Review by noroadsleft — CHANGES_REQUESTED** [review]
GitHub tip: You can apply multiple suggestions to a single commit by using the [Files Changed tab](https://github.com/qmk/qmk_firmware/pull/11246/files).

**Comment by noroadsleft** [comment]
> ### id80.h
> 
> Maps the wiring of the PCB to the physical location on the board. ANSI and ISO have just one main difference here. ISO is having the split left shift in adding pin 11 in the key layout (being KC_NO on ANSI)

Correct.

> ### keyboard.c
> 
> Maps the physical layout of the board to the keymaps. Here ANSI and ISO are not compatible. ANSI is having KC_ENT on pin 83 and KC_BSLS on pin 96. ISO is the other way around as the routing of the wires is different Physically the Enter key belongs to the A-row on ANSi and to the Q-row on ISO.

I think you meant `keymap.c` here, but that's pretty much it.

> If I understand this right, it is not possible to have VIA on this board and have it working out of the box for both variants. You would need to go to VIA first and reassign the key mapping for the enter key. Based on this I don't know if this is a good idea that one of the variants will not work correctly directly after flashing.
> 
> By the way I have a JSON file her ready for a PR for VIA that adds the left split shift key and the different mapping for Enter via two switches.

Actually, come to think of it, I think this will need more changes to properly support VIA. VIA uses the Vendor and Product ID pairings in `config.h` to identify the connected board. As it stands right now, both the ANSI and ISO variants will use the same pairing, but as you've said, the ANSI and ISO keymaps aren't compatible because the PCBs wire the Enter key differently.

The `PRODUCT_ID` and `PRODUCT` strings from `keyboards/id80/config.h` should be moved to `keyboards/id80/ansi/config.h`, and then a `keyboards/id80/iso/config.h` should be created that has a different `PRODUCT_ID` (I'd suggest `0x0081`) and specifies `PRODUCT` as `ID80 ISO`.

**Comment by cwr10010** [comment]
Wait a second.. the changed mapping does not work

**Comment by cwr10010** [comment]
@noroadsleft  As written in my last commit I have a problem now with positioning ENT to the home row. When I compile this with VIA support, the ENT shifts all keys on the Q-row one pin to the right with the supposed ENT key. When I deactivate VIA but have the same keymap definition, the keyboard behaves perfectly. Any idea?

**Comment by noroadsleft** [comment]
@cwr10010 Can you post your VIA JSON file? I'll take a look.

**Comment by cwr10010** [comment]
[id80iso.json.zip](https://github.com/qmk/qmk_firmware/files/5790254/id80iso.json.zip)

Hi James, here is the VIA JSON file. I also have a Gist for the layout editor: http://www.keyboard-layout-editor.com/#/gists/f43b8162d839ceb856d2f8093818d561

Thanks!

**Comment by noroadsleft** [comment]
Hm, the VIA JSON file looks okay. The only things I can think are:

* make sure it's the ISO version's VIA keymap that is flashed to the PCB
  * `make id80/iso:via`, then flash the board – `make id80:via` will compile the ANSI version's VIA keymap instead.
* reset the EEPROM
  1. unplug the board
  2. press and hold the Left Ctrl key
  3. replug the board
  4. release the Left Ctrl key
  5. unplug and replug the board again
* change `KC_BSLS` to `KC_NUHS` in the ISO version's keymaps and reflash the PCB.

Any chance you could post a screenshot of what happens in VIA when you load it on this PCB and keymap? Hard for me to visually/experience as I don't have the board.

**Comment by cwr10010** [comment]
I have reset the EEPROM and this did the trick. Mapping works perfectly now also with the VIA keymap. Thank you!

Is there anything left I should do here now?

**Review by noroadsleft — APPROVED** [review]
Looks good to me.

**Review by drashna — APPROVED** [review]

**Comment by cwr10010** [comment]
Thanks!

**Pull request merged**